File: benchmark/synduce/sorted_list/min.f
Inductive List = elt Int | cons {Int, List};

insert = \y: Int. fix (
  \f: List -> List. \xs: List.
  match xs with
    elt x -> if < y x then cons {y, elt x} else cons {x, elt y}
  | cons {h, t} -> if < y h then cons {y, xs} else cons {h, f t}
  end
);

sort = fix (
  \f: List -> List. \xs: List.
  match xs with
    elt x -> elt x
  | cons {h, t} -> insert h (f t)
  end
);

spec = \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> h
  end;

target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    elt x -> elt x
  | cons {h, t} -> cons {h, f t}
  end
);

main = \xs: List. spec (sort (target xs));

File: ml-benchmark/synduce/sorted_list/min.ml
type l = Elt of int | Cons of int * l

let rec insert y xs =
  match xs with
  | Elt x ->
      if y < x then Cons (y, Elt x) else Cons (x, Elt y)
  | Cons (h, t) ->
      if y < h then Cons (y, xs) else Cons (h, insert y t)

let rec sort xs =
  match xs with
  | Elt x -> Elt x
  | Cons (h, t) -> insert h (sort t)

let spec xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> h

let rec target xs =
  match xs with
  | Elt x -> Elt x
  | Cons (h, t) -> Cons (h, target t)

let program xs = spec (sort (target xs))